eth-contracts/contracts/TokenLocker.sol

```diff
@@ -148,10 +193,129 @@ contract TokenLockerRaw {
         require((recipientCellData.lightClientTypescriptHash() == lightClientTypescriptHash_), "invalid lightClientTypescriptHash in recipient cell");
         require((recipientCellData.bridgeLockscriptCodeHash() == bridgeCellLockscriptCodeHash_), "invalid bridgeLockscriptCodeHash in recipient cell");
         return (
-        recipientCellData.bridgeAmount(),
-        recipientCellData.bridgeFee(),
-        recipientCellData.tokenAddress(),
-        recipientCellData.recipientAddress()
+            recipientCellData.bridgeAmount(),
+            recipientCellData.bridgeFee(),
+            recipientCellData.tokenAddress(),
+            recipientCellData.recipientAddress()
+        );
+    }
+
+    function _getTargetTxRoot(uint64 leafMerkleIndex, TreeNode[] memory leafNodes) internal pure returns (bytes32) {
+        for (uint i = 0; i < leafNodes.length; i++) {
+            if (leafMerkleIndex == leafNodes[i].index) {
+                return leafNodes[i].data;
+            }
+        }
+        return bytes32(0);
+    }
+
+    function _proveTxExist(bytes29 txProofView, bytes32 txHash, bytes32 targetTxRoot)
+    internal
+    view
+    returns (bool)
+    {
+        require(targetTxRoot != bytes32(0), "txRoot from the blockNumber is not in the proof");
+        uint16 index = txProofView.historyTxMerkleIndex();
+        uint16 sibling;
+        uint256 lemmasIndex = 0;
+        bytes29 lemmas = txProofView.historyLemmas();
+        uint256 length = lemmas.len() / 32;
+
+        // calc the rawTransactionsRoot
+        bytes32 rawTxRoot = txHash;
+        while (lemmasIndex < length && index > 0) {
+            sibling = ((index + 1) ^ 1) - 1;
+            if (index < sibling) {
+                rawTxRoot = Blake2b.digest64Merge(rawTxRoot, lemmas.indexH256Array(lemmasIndex));
+            } else {
+                rawTxRoot = Blake2b.digest64Merge(lemmas.indexH256Array(lemmasIndex), rawTxRoot);
+            }
+
+            lemmasIndex++;
+            // index = parent(index)
+            index = (index - 1) >> 1;
+        }
+
+        // calc the transactionsRoot by [rawTransactionsRoot, witnessesRoot]
+        bytes32 transactionsRoot = Blake2b.digest64Merge(rawTxRoot, txProofView.historyWitnessesRoot());
+        require(
+            transactionsRoot == targetTxRoot,
+            "tx proof not passed"
+        );
+        return true;
+    }
+
+    function _proveTxRootExist(bytes29 txRootProofView, bytes32 targetHistoryTxRoot)
+    internal
+    pure
+    returns (TreeNode[] memory leafNodes)
```

Comment:
Does it saves gas if we parse `TreeNode[]` from txRootProofView instead of return it in function?